runtime/compiler/optimizer/EscapeAnalysis.cpp

```diff
@@ -6961,16 +7151,43 @@ void TR_EscapeAnalysis::makeNonContiguousLocalAllocation(Candidate *candidate)
       }
 
    if (candidate->escapesInColdBlocks())
-     {
-     candidate->_originalAllocationNode = candidate->_node->duplicateTree();
-     }
+      {
+      // For newvalue operations, heapification should be performed with a newvalue
+      // that specifies the field values, so the object can be created correctly
+      candidate->_originalAllocationNode = candidate->_node->duplicateTree();
+      }
 
-   // If the object was referenced we will need to create a local object for it
-   // too. In this case, a local object of type "java/lang/Object" is created.
+   // If the object was referenced we will need to create a local object for it too - it's
+   // used to locate all references to the local object at the point of heapification
+   //
+   //   - For TR::New and TR::newvalue, a local object of type "java/lang/Object" is created;
+   //   - For arrays, a local array of length zero is created
    //
    if (candidate->objectIsReferenced())
       {
-      if (candidate->_kind != TR::New)
+      if ((candidate->_kind == TR::New) || (candidate->_kind == TR::newvalue))
+         {
+         // Change the node so that it allocates a java/lang/Object object
+         //
+         TR::ResolvedMethodSymbol *owningMethodSymbol = candidate->_node->getSymbolReference()->getOwningMethodSymbol(comp());
+         TR_OpaqueClassBlock *classObject = comp()->getObjectClassPointer();
+
+         // the call to findOrCreateClassSymbol is safe even though we pass CPI of -1 it is guarded by another check
+         // (see the first occurrence of findOrCreateClassSymbol in EA; dememoizedConstructorCall)
+         //
+         TR::SymbolReference *classSymRef = getSymRefTab()->findOrCreateClassSymbol(owningMethodSymbol, -1, classObject, false);
+         TR::Node *classNode = TR::Node::createWithSymRef(candidate->_node, TR::loadaddr, 0, classSymRef);
+         candidate->_node->removeAllChildren();
+         candidate->_node->setAndIncChild(0, classNode);
+         TR::Node::recreate(candidate->_node, TR::New);
+         candidate->_node->setNumChildren(1);
+         candidate->_class = classObject;
```

Comment:
In making changes to address review comments, I noticed that this change to `candidate->_class` for non-contiguous value type candidates causes problems later in [`heapifyForColdBlocks`](https://github.com/eclipse-openj9/openj9/blob/9e727ec12d64b65b38afe223fec21ab4615810dc/runtime/compiler/optimizer/EscapeAnalysis.cpp#L7500-L7501) when the `_class` field is used to lookup information about the fields of the class.  I will introduce an `_origClass` field to avoid that problem.

Reply:
Fixed in commit fbf9cb620407fe4b1360b4f4603209c253d61535